Services/Administration/OJS.Services.Administration.Business/Implementations/ParticipantScoresBusinessService.cs

```diff
@@ -0,0 +1,127 @@
+namespace OJS.Services.Administration.Business.Implementations;
+
+using FluentExtensions.Extensions;
+using Microsoft.EntityFrameworkCore;
+using OJS.Common.Helpers;
+using OJS.Data.Models.Submissions;
+using OJS.Services.Administration.Data;
+using System.Linq;
+using System.Threading.Tasks;
+
+public class ParticipantScoresBusinessService : IParticipantScoresBusinessService
+{
+    private readonly IParticipantScoresDataService participantScoresData;
+    private readonly IParticipantsDataService participantsData;
+    private readonly ISubmissionsDataService submissionsData;
+
+    public ParticipantScoresBusinessService(
+        IParticipantScoresDataService participantScoresData,
+        IParticipantsDataService participantsData,
+        ISubmissionsDataService submissionsData)
+    {
+        this.participantScoresData = participantScoresData;
+        this.participantsData = participantsData;
+        this.submissionsData = submissionsData;
+    }
+
+    public async Task RecalculateForParticipantByProblem(int participantId, int problemId)
+    {
+        var submission = this.submissionsData.GetBestForParticipantByProblem(participantId, problemId);
+
+        if (submission != null)
+        {
+            await this.participantScoresData.ResetBySubmission(submission);
+        }
+        else
+        {
+            await this.participantScoresData.DeleteForParticipantByProblem(participantId, problemId);
+        }
+    }
+
+    public async Task NormalizeAllPointsThatExceedAllowedLimit()
+    {
+        using var scope = TransactionsHelper.CreateLongRunningTransactionScope();
+        await this.NormalizeSubmissionPoints();
+        await this.NormalizeParticipantScorePoints();
+
+        scope.Complete();
+    }
+
+    private async Task NormalizeSubmissionPoints()
+        => await (await this.submissionsData
+            .GetAllHavingPointsExceedingLimit()
+            .Select(s => new
+            {
+                Submission = s,
+                ProblemMaxPoints = s.Problem!.MaximumPoints,
+            })
+            .ToListAsync())
+            .ForEachSequential(async x =>
+            {
+                x.Submission.Points = x.ProblemMaxPoints;
+
+                this.submissionsData.Update(x.Submission);
+                await this.submissionsData.SaveChanges();
```

Comment:
This will make `N` `SaveChanges()`.
A better way will be to update all submissions in a batch.

Reply:
I made separate issue about this, because the whole method should be rewritten and tested, but it is not used for the functionallities of this PR. It is just copied with the whole service (on later point these services should be cleared all at one, after the whole system is copied to the new project). https://github.com/SoftUni-Internal/exam-systems-issues/issues/125